cpp/src/aztec/CMakeLists.txt

```diff
@@ -88,22 +88,23 @@ if(WASM)
     # Presumably the -O3 when compiling the object files is fine as it's llvms IR optimiser.
     # The backend optimiser is presumably triggered after linking.
     target_link_options(
-        barretenberg-step1.wasm
+        barretenberg.wasm
         PRIVATE
         -nostartfiles -O2 -Wl,--no-entry -Wl,--export-dynamic -Wl,--import-memory -Wl,--allow-undefined -Wl,--stack-first -Wl,-z,stack-size=1048576
     )
 
+    find_program(WASM_OPT wasm-opt)
+    if(NOT WASM_OPT)
+        message(FATAL_ERROR "wasm-opt executable not found. Please install binaryen.")
+    endif()
+
     add_custom_command(
-        OUTPUT ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/barretenberg.wasm
-        COMMAND wasm-opt ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/barretenberg-step1.wasm -O2 --asyncify -o ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/barretenberg.wasm
-        DEPENDS barretenberg-step1.wasm
+        TARGET barretenberg.wasm
+        POST_BUILD
+        COMMAND wasm-opt "$<TARGET_FILE:barretenberg.wasm>" -O2 --asyncify -o "$<TARGET_FILE:barretenberg.wasm>"
         VERBATIM
     )
 
-    add_custom_target(
-        barretenberg.wasm
-        DEPENDS ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/barretenberg.wasm
-    )
```

Comment:
Here we reuse the same target `barretenberg.wasm`. Does this basically say the following?
1. Create an executable target `barretenberg.wasm` that depends on some objects
2. Give that same target some link options
3. Add a custom command for that same target that runs `wasm-opt` (or errors if absent)

Reply:
Sorry for the delay.

You are mostly correct, with slight nuance.
1. Fail if wasm-opt doesn't exist; this now happens during configuration instead of at the end of the build.
2. Create an executable target `barretenberg.wasm` that depends on some objects. (This is the thing that can be installed).
3. Give that same target some link options
4. Add a custom command for that same target that runs wasm-opt on the artifact as a POST_BUILD, so we know the artifact exists and we just optimize and replace the artifact in place. This avoids re-computation if nothing changed and makes installation easier.

Reply:
Ahh okay. So....
1. Since wasm-opt was part of the custom command previously, it wouldnt fail until build. Got it.
2. Got it.
3. Got it.
4. Ah, I didn't notice the `POST_BUILD`. This makes perfect sense, thanks :+1: